src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs

```diff
@@ -0,0 +1,285 @@
+    /// signature. </remarks
+    public Regex ScopedRegex { get; }
+
+    /// <summary>
+    /// A rendering of the signature for fast, allocation-free lookup.
+    /// </summary>
+    /// <remarks>
+    /// This is implemented by packing the 3-4 byte ASCII bytes of the signature
+    /// into an integer, with the leftmost signature character in the least
+    /// signficant byte. However, this encoding should be considered an
+    /// implementation detail that can change.
+    /// </remarks>
+    public int PackedSignature { get; }
+
+    /// <summary>
+    /// Gets the pre-computed pattern for the given signature or null if there is none.
+    /// </summary>
+    public static CompiledHighPerformancePattern? ForSignature(string signature)
+    {
+        return s_patternsBySignature.TryGetValue(signature, out CompiledHighPerformancePattern? pattern) ? pattern : null;
+    }
+
+    /// <summary>
+    /// Regex options used for all high-performance regexes.
+    /// </summary>
+    /// <remarks>
+    /// We don't use 'NonBacktracking' here because it is not supported by the
+    /// regex source generator. However these high-performance patterns should
+    /// not cause problematic backtracking. They are extremely simple, anchored
+    /// to the front with at most one optional group at the end. Generated code
+    /// has been inspected to see that the only "loops" have at most 1
+    /// iteration. If more complex regexes are added, the generated code should
+    /// be re-examined.
+    ///
+    /// 'SingleLine' is used to ensure that '.' matches any character including
+    /// new lines. This is used to share regexes that can purely skip over the
+    /// signature that has already been matched before the regex is applied.
+    ///
+    /// Also note that the high-performance scanner does not allow substitution
+    /// of the regex engine and the regex engine built-in to .NET is always
+    /// used. Profiling showed no performance gain from RE2, even on .NET
+    /// Framework, even for the unoptimized 'SecretMasker' using only the
+    /// highly-identifiable patterns, but this merits more scrutiny and
+    /// investigation.
+    ///
+    /// Even more critically, the regex engine abstraction does not currently
+    /// provide a mechanism to search within a substring without allocation.
+    /// There is no incentive to add it, which would require work in the RE2
+    /// wrapper, unless further investigation reveals a flaw in the finding on
+    /// RE2 performance above.
+    /// </remarks>
+    private const RegexOptions Options = RegexOptions.Compiled |
+                                         RegexOptions.ExplicitCapture |
+                                         RegexOptions.CultureInvariant |
+                                         RegexOptions.Singleline;
+
+
+    public CompiledHighPerformancePattern(string signature,
+                                           int signaturePrefixLength,
```

Comment:
>int [](http://example.com/codeflow?start=43&length=3)

NIT parameters not in alignment

Reply:
Because of refactor -> rename, no doubt. 😅